pageserver/src/tenant/delete.rs

```diff
@@ -0,0 +1,580 @@
+                None, // Ancestor is not needed for deletion.
+                remote_client,
+                init_order,
+                // Important. We dont pass ancestor above because it can be missing.
+                // Thus we need to skip the validation here.
+                false,
+            )
+            .context("create_timeline_struct")?;
+
+        let mut guard = DeletionGuard(
+            Arc::clone(&timeline.delete_progress)
+                .try_lock_owned()
+                .expect("cannot happen because we're the only owner"),
+        );
+
+        // Note: here we even skip populating layer map. Timeline is essentially uninitialized.
+        // RemoteTimelineClient is the only functioning part.
+        timeline.set_state(TimelineState::Stopping);
+        // We meed to do this because when console retries delete request we shouldnt answer with 404
+        // because 404 means successful deletion.
+        {
+            let mut locked = tenant.timelines.lock().unwrap();
+            locked.insert(timeline_id, Arc::clone(&timeline));
+        }
+
+        guard.mark_in_progress()?;
+
+        Self::schedule_background(guard, tenant.conf, tenant, timeline);
+
+        Ok(())
+    }
+
+    pub async fn cleanup_remaining_fs_traces_after_timeline_deletion(
+        tenant: &Tenant,
+        timeline_id: TimelineId,
+    ) -> anyhow::Result<()> {
+        cleanup_remaining_fs_traces_after_timeline_deletion(
+            tenant.conf,
+            tenant.tenant_id,
+            timeline_id,
+        )
+        .await
+    }
+
+    fn prepare(
+        tenant: &Tenant,
+        timeline_id: TimelineId,
+    ) -> Result<(Arc<Timeline>, DeletionGuard), DeleteTimelineError> {
+        let timelines = tenant.timelines.lock().unwrap();
+
+        let timeline = match timelines.get(&timeline_id) {
+            Some(t) => t,
+            None => return Err(DeleteTimelineError::NotFound),
+        };
+
+        // Ensure that there are no child timelines **attached to that pageserver**,
+        // because detach removes files, which will break child branches
+        let children: Vec<TimelineId> = timelines
+            .iter()
+            .filter_map(|(id, entry)| {
+                if entry.get_ancestor_timeline_id() == Some(timeline_id) {
+                    Some(*id)
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        if !children.is_empty() {
+            return Err(DeleteTimelineError::HasChildren(children));
+        }
+
+        let delete_lock_guard = DeletionGuard(
+            Arc::clone(&timeline.delete_progress)
+                .try_lock_owned()
+                .map_err(|_| DeleteTimelineError::AlreadyInProgress)?,
+        );
+
+        timeline.set_state(TimelineState::Stopping);
+
+        Ok((Arc::clone(timeline), delete_lock_guard))
+    }
+
+    pub fn is_deleted(&self) -> bool {
+        matches!(self, Self::Finished)
+    }
```

Comment:
Not used. Should it be? I guess there's a possible race:

```
T1: get preempted right before the try_lock on Timeline::deletion
T2: do a full deletion, acquire and drop Timeline::deletion lock
T1: acquire deletion lock, do another `DeleteTimelineFlow::run`
```

The behavior is probably fine (I guess we error out in line 3), but, it would be nice to detect this condition early in `DeleteTimelineFlow`, by checking for the current state, and erroring out if it's NotStarted.

We won't need this method then.

Reply:
See https://github.com/neondatabase/neon/pull/4619#discussion_r1251777216. Does it make sense?

Reply:
I think this method just needs to be removed

Reply:
>  When we finish the deletion we remove timeline from the map first and then drop the guard, so I dont see how we can get a timeline with Finished guard from the map

We don't check for `Finished` in the `DeleteTimelineFlow::prepare()`, do we?

So, my race above can still happen? Both T1 and T2 observed an `Active` timeline when they started.

Reply:
Removed the method

Reply:
@problame should this one be resolved? Or you still see the race here?

Reply:
@problame ping

Reply:
Ah ok, got it. 

If `T1` gets preempted right before here:

```
 let delete_lock_guard = DeletionGuard(
            Arc::clone(&timeline.delete_progress)
                .try_lock_owned()
                .map_err(|_| DeleteTimelineError::AlreadyInProgress)?,
        );
```

and then T2 runs. 

Then T1 resumes, it will win the lock and get a guard.

But then

```
        guard.mark_in_progress()?;
```

gets called

```
fn mark_in_progress(&mut self) -> anyhow::Result<()> {
        match self {
            Self::Finished => anyhow::bail!("Bug. Is in finished state"),
            Self::InProgress { .. } => { /* We're in a retry */ }
            Self::NotStarted => { /* Fresh start */ }
        }

        *self = Self::InProgress;

        Ok(())
    }
```

and we will fail `prepare()` with `anyhow::bail!("Bug. Is in finished state")`.

So, we're detecting the condition before T1 tries to change anything, which would be futile, because all state is already gone.

The  error messages could be more descriptive, though. 


I guess feel free to close as my concern is eliminated, but I think it's totally non-obvious that the `guard.mark_in_progress()` protects us that way.

Can we have a more descriptive error message, or at least a comment?

Basically, it needs to be exaplained that the `Arc<Timeline>` can live longer than its entry in the `Tenant::timelines` map if there are concurrent callers. Example: see above.

Reply:
I still dont understand how that race is possible.

> If T1 gets preempted right before here:
> and then T2 runs.

If I understand it correctly T2 will wait until T1 releases the lock on timelines hashmap, and when the lock is released deletion guard is taken, so it should return an error.

Am I missing something?

I think at some point I had a comment on importance of ordering timeline removal from the map before dropping the guard, cant find it now, will add.

Reply:
Discussed verbally, this race is not possible because we hold the lock on timelines map while acquiring a lock on deletion guard. 

I'll add a comment explaining this (and few other aspects) and resolve the discussion after that

Reply:
Pushed https://github.com/neondatabase/neon/pull/4552/commits/9fc19f96152ee92740733c4ea68a8da670050720

Resolving, please reopen if you want me to change the comment